### PR TITLE
Fixes Firebase Collection child changed

### LINKED ISSUE #19: firebase-collection: Uncaught TypeError: Cannot read property 'slice' of undefined
Using firebase-collection this way:
```html
<!doctype html>
<html lang="">
  <head>
    <meta charset="utf-8">
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Polymer Starter Kit</title>
    
    <script src="bower_components/webcomponentsjs/webcomponents-lite.js"></script>
  
    <link rel="import" href="bower_components/firebase-element/firebase-collection.html">
  </head>
  <body>
    <template is="dom-bind">
      <firebase-collection
        location="https://example.firebaseio.com/entries"
            data="{{entries}}"></firebase-collection>
          <template is="dom-repeat" items="[[entries]]" as="entry">
            <span>[[entry.name]]</span>
          </template>
    </template>
  </body>
</html>
```

Works fine on first run, but when updating it shows following error in the console:
```
firebase-collection: Uncaught TypeError: Cannot read property 'slice' of undefined
```

### PR DESCRIPTION
**Pull request by cdata**
- Fixes an incorrect usage of `Polymer.Base.splice`
 - Adds reflection of local changes to remote children
 - Adds a test for changing children

Fixes #19 

/cc @notwaldorf @morethanreal

### CODE REVIEW

File: firebase-collection.html [hunk 1]
```diff
@@ -268,6 +268,20 @@
       this.query.ref().child(key).remove();
     },
 
+    _applyLocalDataChanges: function(change) {
+      var pathParts = change.path.split('.');
+      var index;
```
Comment (notwaldorf) [inline firebase-collection.html]: maybe move the declarations to L280/281 where they're actually needed?

File: firebase-collection.html [hunk 2]
```diff
@@ -268,6 +268,20 @@
...
+      var value;
+
+      if (pathParts.length < 2 || pathParts[1] === 'splices') {
+        return;
+      }
+
+      index = window.parseInt(pathParts[1], 10);
```
Comment (notwaldorf) [inline firebase-collection.html]: do you need `window.`? (you probably do, I just haven't seen that before)

File: test/firebase-collection.html [hunk 3]
```diff
@@ -40,6 +40,15 @@
       </firebase-collection>
     </template>
   </test-fixture>
+  <test-fixture id="ChangingChildren">
```
Comment (notwaldorf) [inline test/firebase-collection.html]: 👶

File: test/firebase-collection.html [hunk 4]
```diff
@@ -148,6 +157,55 @@
...
+            done(e)
+          });
+        });
+
+        test('updates the child key in place with the new value', function(done) {
+          var childrenKeys = [];
+
+          waitForEvent(firebase, 'firebase-value').then(function() {
+            var middleValue = firebase.getByKey(childrenKeys[1]);
+            var changes;
```
Comment (notwaldorf) [inline test/firebase-collection.html]: maybe move `changes` to L180?

File: test/firebase-collection.html [hunk 5]
```diff
@@ -148,6 +157,55 @@
         });
       });
 
+      suite('a child changes', function() {
```
Comment (notwaldorf) [inline test/firebase-collection.html]: this is almost existential. i like it.

**Comment by notwaldorf** [comment]
minor comments but :+1:

File: firebase-collection.html (hunk 1 above)
Comment (cdata) [inline firebase-collection.html]: This is bike-sheddy of me, but I like to keep the declarations where they will be hoisted.

File: test/firebase-collection.html (hunk 5 above)
Comment (addyosmani) [inline test/firebase-collection.html]: :baby: :boy:

File: firebase-collection.html (hunk 2 above)
Comment (addyosmani) [inline firebase-collection.html]: I've never seen it used via the `window` object before, but one possible explanation is that directly calling window can reduce cost of the scope chain looking for any other definitions of `parseInt`. Just a guess.

File: firebase-collection.html (hunk 1 above)
Comment (addyosmani) [inline firebase-collection.html]: sgtm

**Comment by addyosmani** [comment]
Your call on https://github.com/GoogleWebComponents/firebase-element/pull/26#discussion-diff-31592801, but changes otherwise LGTM. Nice one!

File: firebase-collection.html (hunk 2 above)
Comment (cdata) [inline firebase-collection.html]: @addyosmani is being generous. It's just an old habit. I'm happy to remove.

File: test/firebase-collection.html (hunk 4 above)
Comment (cdata) [inline test/firebase-collection.html]: Ditto, see https://github.com/GoogleWebComponents/firebase-element/pull/26#discussion_r31592722.

**Pull request merged**